includes/class-llms-order-generator.php

```diff
@@ -0,0 +1,732 @@
+	 */
+	protected $student = null;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $data {
+	 *     An associative array of input data used to generate the order, usually from $_POST.
+	 *
+	 *     @type integer $llms_plan_id         An LLMS_Access_Plan ID.
+	 *     @type string  $llms_agree_to_terms  A yes/no value determining whether or not the user has agreed to the site's terms.
+	 *     @type string  $llms_payment_gateway The ID of the payment gateway used to process the order.
+	 *     @type string  $llms_coupon_code     Optional. The coupon code string being used.
+	 *     @type string  $llms_order_key       Optional. An `LLMS_Order` key used to modify an existing pending order rather than creating a new one.
+	 *     @type array   ...$user_data         All remaining data is passed to the user creation functions.
+	 * }
+	 */
+	public function __construct( $data ) {
+		$this->data = $data;
+	}
+
+	/**
+	 * Confirms an existing pending order.
+	 *
+	 * @since [version]
+	 *
+	 * @return WP_Error|array Returns an array of data from the payment gateway's `confirm_pending_order()` method on success.
+	 */
+	public function confirm() {
+
+		$validate = $this->validate( true );
+		if ( is_wp_error( $validate ) ) {
+			return $validate;
+		}
+
+		$gateway_confirm = $this->gateway->confirm_pending_order( $this->order );
+		if ( is_wp_error( $gateway_confirm ) ) {
+			return $gateway_confirm;
+		}
+
+		$user = $this->commit_user();
+		if ( is_wp_error( $user ) ) {
+			return $user;
+		}
+
+		// Save the user to the order.
+		$this->order->set_user_data( $this->get_user_data() );
+
+		if ( 'SUCCESS' === ( $gateway_confirm['status'] ?? null ) && ! empty( $gateway_confirm['transaction'] ) ) {
+			// Record the transaction.
+			$this->order->record_transaction( $gateway_confirm['transaction'] );
+		}
+
+		return $gateway_confirm;
+
+	}
+
+
+	/**
+	 * Creates a new pending order.
+	 *
+	 * @since [version]
+	 *
+	 * @return WP_Error|LLMS_Order
+	 */
+	protected function create() {
+
+		$order = new LLMS_Order( $this->get_order_id() );
+
+		// If there's no id we can't proceed, return an error.
+		if ( ! $order->get( 'id' ) ) {
+			return $this->error(
+				self::E_CREATE_ORDER,
+				__( 'There was an error creating your order, please try again.', 'lifterlms' )
+			);
+		}
+
+		$order->init( $this->get_user_data(), $this->plan, $this->gateway, $this->coupon );
+
+		return $order;
+
+	}
+
+	/**
+	 * Registers or updates the user from the submitted data.
+	 *
+	 * @since [version]
+	 *
+	 * @return integer|WP_Error Returns the `WP_User` ID on success or an error object.
+	 */
+	protected function commit_user() {
+
+		$args = array(
+			'plan' => $this->plan,
+		);
+
+		$user = get_current_user_id() ?
+			llms_update_user( $this->data, 'checkout', $args ) :
+			llms_register_user( $this->data, 'checkout', true, $args );
+
+		if ( ! is_wp_error( $user ) ) {
+			$this->student = llms_get_student( $user );
```

Comment:
we cache the student here, why? Can't we just cache the WP_User?
Where are we going to use the student object?

Reply:
Because it's LifterLMS and the `get_student()` method returns a student object.

Reply:
I know, but it's less performant :P